### PR TITLE
Update client to TypeScript 5.3.3

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Typescript 5.3 brings useful features like support for [Symbol.hasInstance narrowing](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-3.html#instanceof-narrowing-through-symbolhasinstance), type checking optimizations and much more. Since ClientRequirements.md lists our requirement at TypeScript 5.4, this should be ok (from the perspective of aligning with that requirement)

See https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-2.html and https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-3.html for more details.

Based on a two test runs, this change makes `pnpm run build:compile` about 7% faster. (105.438 and 107.011 seconds instead of 113.816 and 114.096 without the change)

## Breaking Changes

This is not expected to break anything, however its possible that the compile TypeScript is slightly different, which could cause issues in some edge cases.

This will enable code to start relying on newer features, which when used in public APIs could in turn cause breakages for users of older versions of TypeScript.

There are a couple cases where type-tests were impacted. These seemed to be a fix of typescript capturing more accurate types, especially around index signatures, and cases where the TypeScript compiler wouldn't provide any useful information about what changed with the only difference being inside the `...` non-printed parts of the types. As the changes seem relatively minor they seem unlikely to cause any real issues (didn't break lots of examples/tests) and since we only support customers using TypeScript 5.4 and plan to move to 5.4 ourselves at some point, if this causes a breaking change in some edge case, that probably a good thing to get out of the way now while we are still in the RC phase.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

Does this cause issues with Fluid-Build's incremental build logic that needs fixing?
Is there anything else that should be validated?

When running build:compile with no changes needing building, it used to run 417 tasks, and now it runs 425. The reason for the difference is unclear, but it still takes 16 seconds (Updated version was 15.5 seconds, old one 16.2). The extra tasks are

> [  1/425] ✓ @fluid-experimental/property-properties: tsc --project ./tsconfig.esnext.json - 0.449s
> [  2/425] ✓ @fluid-experimental/property-properties: tsc - 0.490s
> [  3/425] - @fluid-experimental/property-proxy: fluid-tsc commonjs --project ./tsconfig.cjs.json - 0.007s
> [  4/425] - @fluid-experimental/property-proxy: tsc --project ./tsconfig.json - 0.009s
> [  5/425] - @fluid-experimental/property-proxy: copyfiles -f ../../../../common/build/build-common/src/cjs/package.json ./dist - 0.001s
> [ 21/425] ✓ @fluid-experimental/property-proxy: tsc --project ./src/test/tsconfig.json - 0.537s
> [ 22/425] ✓ @fluid-experimental/property-properties: tsc --project ./src/test/tsconfig.json - 0.718s
> [ 23/425] ✓ @fluid-experimental/property-proxy: fluid-tsc commonjs --project ./src/test/tsconfig.cjs.json - 0.695s

An explanation of why this change occurred might be worth understanding.

### CODE REVIEW

File: examples/data-objects/monaco/webpack.config.cjs [hunk 1]
```diff
@@ -4,107 +4,68 @@
...
-			module: {
-				rules: [
-					{
-						test: /\.tsx?$/,
-						use: [
-							{
-								loader: "ts-loader",
-								options: {
-									compilerOptions: {
-										module: "esnext",
```
Comment (CraigMacomber) [inline examples/data-objects/monaco/webpack.config.cjs:29]: This config caused ` TS5110: Option 'module' must be set to 'Node16' when option 'moduleResolution' is set to 'Node16'.` So I made it mased on our default config. The dependencies still cause source maps issues which make the app not work when manually testing it), so I had to keep source map suppression.

File: packages/dds/sequence/api-report/sequence.api.md [hunk 2]
```diff
@@ -317,7 +317,7 @@ export type IntervalStickiness = (typeof IntervalStickiness)[keyof typeof Interv
 export enum IntervalType {
     // (undocumented)
     Simple = 0,
-    SlideOnRemove = 2,
+    SlideOnRemove = 2,// SlideOnRemove is default behavior - all intervals are SlideOnRemove
```
Comment (CraigMacomber) [inline packages/dds/sequence/api-report/sequence.api.md:320]: New typescript seems to result in preserved trailing comments in some cases like this. Interesting, but I think a non-issue.

File: packages/framework/fluid-static/package.json [hunk 3]
```diff
@@ -140,6 +140,34 @@
 			"TypeAliasDeclaration_Myself": {
 				"forwardCompat": false,
 				"backCompat": false
+			},
+			"InterfaceDeclaration_IFluidContainer": {
```
Comment (CraigMacomber) [inline packages/framework/fluid-static/package.json:144]: The API breakages seem to be (for the ones I looked at in detail anyway) improved type safety catching more incompatibilities from the IFLuidHandle refactor which were previously missed.

File: packages/dds/tree/src/test/feature-libraries/optional-field/optionalFieldUtils.ts [hunk 4]
```diff
@@ -152,7 +152,7 @@ export const Change = {
 		const changeset: Mutable<OptionalChangeset> = { moves, childChanges };
 		for (const changeLike of changes) {
 			if ("type" in changeLike === false) {
-				const change = changeLike as OptionalChangeset;
+				const change = changeLike;
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/feature-libraries/optional-field/optionalFieldUtils.ts:155]: improved narrowing made this `as` cast a no-op, so it had to be removed as per our compiler settings.

File: packages/drivers/odsp-driver/src/odspDelayLoadedDeltaStream.ts [hunk 5]
```diff
@@ -147,7 +147,7 @@ export class OdspDelayLoadedDeltaStream {
 			const websocketTokenPromise = requestWebsocketTokenFromJoinSession
 				? // eslint-disable-next-line unicorn/no-null
 				  Promise.resolve(null)
-				: this.getWebsocketToken!(options);
+				: this.getWebsocketToken(options);
```
Comment (CraigMacomber) [inline packages/drivers/odsp-driver/src/odspDelayLoadedDeltaStream.ts:150]: improved narrowing removed the new for this non-null assertion which thus had to be removed due to our compiler settings.

**Comment by tylerbutler** [comment]
What happens when someone uses TS 5.1? How does it break? How do they know their ts version is the problem?

File: packages/dds/sequence/api-report/sequence.api.md (hunk 2 above)
Comment (tylerbutler) [reply packages/dds/sequence/api-report/sequence.api.md:320]: I seem to remember @DLehenbauer mentioning that we should try to upgrade api-extractor and typescript at the same time and I think these sorts of things was one reason. Though @Josmithr has also updated api-extractor independently so I could be way off here.

**Review by ChumpChief — COMMENTED** [review]
⁠We talked about this in API council yesterday - the only concern raised was whether customers on older versions of TS might have trouble consuming .d.ts files we make using newer TS (even if not on day 0, maybe later as we start taking dependencies on new features).

Have you done research on current versions of TS used by our customers and/or do we know what usage distribution looks like for the various TS versions?

I'm in favor of this change on a personal level, but would be good to have confidence that it won't become disruptive to customers.

File: packages/dds/sequence/api-report/sequence.api.md (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/sequence/api-report/sequence.api.md:320]: We are already using a version of API extractor which uses newer typescript than we build with. This change brings them in alignment again (if our typescript is newer than theirs, we get warnings)

**Comment by Josmithr** [comment]
> ⁠We talked about this in API council yesterday - the only concern raised was whether customers on older versions of TS might have trouble consuming .d.ts files we make using newer TS (even if not on day 0, maybe later as we start taking dependencies on new features).
> 
> Have you done research on current versions of TS used by our customers and/or do we know what usage distribution looks like for the various TS versions?
> 
> I'm in favor of this change on a personal level, but would be good to have confidence that it won't become disruptive to customers.

We do now have documentation noting that `TS 5.4` is a minimal requirement. I would argue that the time to do that sort of investigation is when updating our min requirements, rather than when updating dependencies in accordance with those requirements.

If we want to retain compatibility with earlier versions of TS and avoid consumer disruption for users upgrading to FF2.0, we will need to update our documented min. But I also think we need to be somewhat aggressive when increment our major versions, otherwise we will be stuck on very old versions for a long time.

**Comment by CraigMacomber** [comment]
> Have you done research on current versions of TS used by our customers and/or do we know what usage distribution looks like for the various TS versions?

I had Nick run our ClientRequirements.md document (which says we require 5.4 from our customers) by various people for review and got no pushback on that. Given that we already require TS > 5, and there aren't and major compat issues with updating TS from 5.1 to 5.3 that I'm aware of I don't expect this to cause significant issues.

I imagine these stats are not super useful, but in the last 7 days npm shows almost 5 million downloads of the 5.3.3 version this targets, so its the most popular version in the 5.0 - 5.3 range by a significant margin: not many people that bothered getting onto 5 stopped at 5.1 or 5.2. 5.4.5 is the most popular version (over 9 million downloads).

api-extractor moved to ts 5.3.3 5 months ago in 7.39, and that was the first version of it that moved past 5.0, so there are no api-extractor releases supporting our current 5.1 that don't support 5.3.

**Comment by ChumpChief** [comment]
> I would argue that the time to do that sort of investigation is when updating our min requirements, rather than when updating dependencies in accordance with those requirements.

I agree with this, so if the investigation was done that's great - I just haven't seen it.

My point isn't to say "don't move forward if it would disrupt" - my point is to say "we should know how disruptive it will be if we move forward".  So that whatever decision we make can be informed by customer impact.

**Review by nmsimons — APPROVED** [review]
We've vetted this requirement with Hexagon and Autodesk as well as internal SharePoint pages. Consensus was that this is a reasonable requirement.

**Review by Josmithr — APPROVED** [review]

File: examples/data-objects/monaco/webpack.config.cjs [hunk 6]
```diff
@@ -4,107 +4,68 @@
...
+			},
+		},
+		output: {
+			chunkFilename: "[name].async.js",
+			publicPath: "/dist/",
+			globalObject: "self",
+		},
+		plugins: [
+			new MonacoWebpackPlugin(),
+			// new BundleAnalyzerPlugin()
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply examples/data-objects/monaco/webpack.config.cjs]: Package.json does not even depend on the dep needed to provide this, so ya, seems good to remove it. Done.

**Pull request merged**